packages/rest/lib/RestManager.ts

```diff
@@ -0,0 +1,214 @@
+import fetch from "node-fetch";
+import { ENDPOINTS } from "./constants";
+import {
+    RestOptions,
+    RequestMethods,
+    MessageContent,
+    MessagePayload,
+    GetChannelMessagesOptions,
+    UpdateChannelMessageOptions,
+    CreateForumThreadOptions,
+    ForumThreadPayload,
+    ListItemPayload,
+    CreateListItemOptions,
+    CreateDocOptions,
+    DocPayload,
+    UpdateDocOptions,
+    MemberXPPayload,
+    SocialLinkType,
+} from "./typings";
+
+export class RestManager {
+    /** The bot token to be used for making requests. */
+    token: string;
+    /** The version of the API to be used for making requests. By default, this will use the latest version that the library supports. */
+    version: number;
+    /** The base of the URL that this will send this request to. By default, this will use guilded's api URL. */
+    baseURL: string;
+    /** The proxy url if it was set. */
+    proxyURL: string;
+
+    constructor(options: RestOptions) {
+        this.token = options.token;
+        this.version = options.version || 1;
+        this.baseURL = "https://www.guilded.gg/api";
+        this.proxyURL = options.proxyURL || "";
+    }
+
+    /** Combine the base url and the query to get the final url to send to. */
+    getFinalURL(query: string) {
+        return `${this.url}/${query}`;
+    }
+
+    /** The full url to send the request to. */
+    get url() {
+        return this.proxyURL || `${this.baseURL}/v${this.version}`;
+    }
+
+    /** Make a POST request to the API. */
+    async post<T>(url: string, body: Record<string, unknown>): Promise<T> {
+        return this.sendRequest("POST", url, body);
+    }
+
+    /** Make a GET request to the API. */
+    async get<T>(url: string, body?: Record<string, string>): Promise<T> {
+        return this.sendRequest("GET", `${url}/${new URLSearchParams(body).toString()}`, {});
+    }
+
+    /** Make a PUT request to the API. */
+    async put<T>(url: string, body?: Record<string, unknown>): Promise<T> {
+        return this.sendRequest("PUT", url, body ?? {});
+    }
+
+    /** Make a DELETE request to the API. */
+    async delete<T>(url: string, body?: Record<string, unknown>): Promise<T> {
+        return this.sendRequest("DELETE", url, body || {});
+    }
+
+    /** Make a PATCH request to the API. */
+    async patch<T>(url: string, body: Record<string, unknown>): Promise<T> {
+        return this.sendRequest("PATCH", url, body);
+    }
+
+    /** Send a request to the api with all the data provided. */
+    async sendRequest<T>(method: RequestMethods, url: string, body: Record<string, unknown>): Promise<T> {
+        return fetch(url, {
+            method,
+            body: JSON.stringify(body),
+            headers: {
```

Comment:
Get does not have a body according to spec.

```suggestion
    async get<T>(url: string): Promise<T> {
        return this.sendRequest("GET", `${url}/${new URLSearchParams(body).toString()}`);
    }

    /** Make a PUT request to the API. */
    async put<T>(url: string, body?: Record<string, unknown>): Promise<T> {
        return this.sendRequest("PUT", url, body);
    }

    /** Make a DELETE request to the API. */
    async delete<T>(url: string, body?: Record<string, unknown>): Promise<T> {
        return this.sendRequest("DELETE", url, body);
    }

    /** Make a PATCH request to the API. */
    async patch<T>(url: string, body: Record<string, unknown>): Promise<T> {
        return this.sendRequest("PATCH", url, body);
    }

    /** Send a request to the api with all the data provided. */
    async sendRequest<T>(method: RequestMethods, url: string, body?: Record<string, unknown>): Promise<T> {
        return fetch(url, {
            method,
            body: body && JSON.stringify(body),
            headers: {
```